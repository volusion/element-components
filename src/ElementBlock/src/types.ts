export interface ElementBlockStylesProps {
    isFullWidth: boolean;
    hasHorizontalPadding: boolean;
    topPadding: paddingSizes;
    bottomPadding: paddingSizes;
    showOnDesktop: boolean;
    showOnTablet: boolean;
    showOnMobile: boolean;
}

type paddingSizes = 'None' | 'Small' | 'Medium' | 'Large';

export interface ElementBlockConfigs extends ElementBlockStylesProps {
    customAttrs: customAttrsProps[];
}

export interface ElementBlockProps extends ElementBlockConfigs {
    children: any;
    globalSettings: any;
}

export interface customAttrsProps {
    name: string;
    value: string;
}
export interface ObjectLiteral {
    [key: string]: any;
}
