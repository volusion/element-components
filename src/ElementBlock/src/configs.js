import { ElementPropTypes } from '@volusion/element-proptypes';

export const configSchema = {
    text: {
        label: 'Text content',
        type: ElementPropTypes.string
    }
};

export const defaultConfig = {
    text: 'Element Starter Block'
};
