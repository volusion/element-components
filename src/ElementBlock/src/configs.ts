import { ElementBlockConfigs } from './types';
import { ElementPropTypes } from '@volusion/element-proptypes';

export const configSchema = {
    isFullWidth: {
        label: `Full width`,
        type: ElementPropTypes.bool
    },
    hasHorizontalPadding: {
        label: `Horizontal spacing`,
        type: ElementPropTypes.bool
    },
    topPadding: {
        label: `Top spacing`,
        type: ElementPropTypes.oneOf(['None', 'Small', 'Medium', 'Large'])
    },
    bottomPadding: {
        label: `Top spacing`,
        type: ElementPropTypes.oneOf(['None', 'Small', 'Medium', 'Large'])
    },
    showOnDesktop: {
        label: `Show on desktop`,
        type: ElementPropTypes.bool
    },
    showOnTablet: {
        label: `Show on tablet`,
        type: ElementPropTypes.bool
    },
    showOnMobile: {
        label: `Show on mobile`,
        type: ElementPropTypes.bool
    },
    customAttrs: {
        label: `Custom element attributes`,
        type: ElementPropTypes.arrayOf(
            ElementPropTypes.shape({
                name: {
                    label: `Name`,
                    type: ElementPropTypes.string
                },
                value: {
                    label: `Value`,
                    type: ElementPropTypes.string
                }
            })
        )
    }
};

export const defaultConfig: ElementBlockConfigs = {
    isFullWidth: true,
    hasHorizontalPadding: false,
    topPadding: 'None',
    bottomPadding: 'None',
    showOnDesktop: true,
    showOnTablet: true,
    showOnMobile: true,
    customAttrs: [
        {
            name: '',
            value: ''
        }
    ]
};
