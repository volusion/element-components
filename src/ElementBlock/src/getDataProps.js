export const getDataProps = (utils, props) => Promise.resolve({});
