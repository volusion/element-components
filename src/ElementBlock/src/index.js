export { default as block } from './Block';
export { configSchema, defaultConfig } from './configs';
export { getDataProps } from './getDataProps';
