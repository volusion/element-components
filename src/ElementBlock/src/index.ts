export { default as block } from './Block';
export { configSchema, defaultConfig } from './configs';
