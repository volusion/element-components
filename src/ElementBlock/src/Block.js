import React from 'react';
import { css, StyleSheet } from 'aphrodite';
import { getStyles } from './getStyles';

import { defaultConfig } from './configs';

const Block = props => {
    const classes = StyleSheet.create(getStyles(props));
    return <h1 className={css(classes.example)}>{props.text}</h1>;
};

Block.defaultProps = defaultConfig;

export default Block;
