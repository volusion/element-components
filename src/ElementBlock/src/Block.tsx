import React from 'react';
import { css, StyleSheet } from 'aphrodite';
import { getStyles } from './getStyles';
import { defaultConfig } from './configs';
import { ElementBlockProps, ObjectLiteral } from './types';

const Block = (props: ElementBlockProps) => {
    const { globalSettings, customAttrs, children } = props;
    if (!children) return null;
    const styles = StyleSheet.create(getStyles(props, globalSettings));
    const customProps: ObjectLiteral = {};
    customAttrs.filter(attr => attr.name).forEach((attr: { name: string; value: string }) => {
        const hyphenateString = (string = '') => {
            const specialChars = /[^a-zA-Z0-9\s-_]+/g;
            const nonAlphaNumericChars = /[^a-zA-Z0-9]+/g;
            const attr = string
                .trim()
                .replace(specialChars, '')
                .replace(nonAlphaNumericChars, '-')
                .toLowerCase()
            return attr;
        };
        const fullAttrName = `data-element-${hyphenateString(attr.name)}`;
        customProps[fullAttrName] = attr.value;
    });
    return (
        <div className={css(styles.wrapper)} {...customProps}>
            {children}
        </div>
    );
};

Block.defaultProps = defaultConfig;

export default Block;
