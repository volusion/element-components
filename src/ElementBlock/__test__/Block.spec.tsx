import React from 'react';
import { StyleSheetTestUtils } from 'aphrodite';
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });
import { block as Block } from '../src/';
import { defaultConfig } from '../src/configs';
import { ElementBlockProps } from '../src/types';

let props: ElementBlockProps = {
    ...defaultConfig,
    children: 'TEST_CHILD',
    globalSettings: {}
};
describe('The ElementBlock component', () => {
    StyleSheetTestUtils.suppressStyleInjection();

    it('should match the snapshot', () => {
        const wrapper = mount(<Block {...props} />);
        expect(wrapper.render()).toMatchSnapshot();
    });
    describe('when there are no children', () => {
        it('should return null', () => {
            const wrapper = mount(<Block {...props} children={undefined} />);
            expect(wrapper.isEmptyRender()).toBeTruthy();
        });
    });
    describe('when the custom attribute does not have a valid name', () => {
        it('should not apply the attribute to the element', () => {
            const customAttrs = [{ name: '', value: 'TEST_VALUE' }];
            const wrapper = mount(<Block {...props} customAttrs={customAttrs} />);
            expect(Object.keys(wrapper.find('div').props())).toHaveLength(2); // ["className", "children"]
        });
    });
    describe('when the custom attribute has a valid name', () => {
        it('should apply the attribute to the element', () => {
            const customAttrs = [{ name: 'TEST_NAME', value: 'TEST_VALUE' }];
            const wrapper = mount(<Block {...props} customAttrs={customAttrs} />);
            expect(wrapper.exists(`[data-element-test-name="TEST_VALUE"]`)).toBeTruthy();
        });
    });
});
