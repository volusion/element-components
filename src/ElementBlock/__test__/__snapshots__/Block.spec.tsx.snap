// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The ElementBlock component should match the snapshot 1`] = `
<div
  class="wrapper_zdxht7"
>
  TEST_CHILD
</div>
`;
